src/util/subscription-set/expiring-sorted-set.ts

```diff
@@ -16,28 +19,83 @@ interface ExpiringSortedSetEntry<T> {
  * @typeParam T - the type of the set entries' values
  */
 export class ExpiringSortedSet<T> implements SubscriptionSet<T> {
-  map = new Map<string, ExpiringSortedSetEntry<T>>()
+  capacity: number
+  map: Map<string, LinkedListNode<ExpiringSortedSetEntry<T>>>
+  list: DoubleLinkedList
+
+  constructor(capacity: number) {
+    this.capacity = capacity
+    this.map = new Map<string, LinkedListNode<ExpiringSortedSetEntry<T>>>()
+    this.list = new DoubleLinkedList()
+  }
 
   add(key: string, value: T, ttl: number) {
-    this.map.set(key, {
-      value,
-      expirationTimestamp: Date.now() + ttl,
-    })
+    if (this.map.has(key)) {
+      const node = this.map.get(key) as LinkedListNode<ExpiringSortedSetEntry<T>>
+
+      node.data = {
+        value,
+        expirationTimestamp: Date.now() + ttl,
+      }
+      this.moveToTail(node)
+    } else {
+      this.evictIfNeeded()
+      const data = {
+        value,
+        expirationTimestamp: Date.now() + ttl,
+      }
+      const node = new LinkedListNode(key, data)
+      this.list.insertAtTail(node)
+      this.map.set(key, node)
+    }
   }
 
-  getAll(): T[] {
+  getAll(): PromiseOrValue<T[]> {
     const results: T[] = []
-    const now = Date.now()
+    for (const [key] of this.map.entries()) {
+      const value = this.get(key) as T
+      if (value) {
+        results.push(value)
+      }
+    }
+    return results
+  }
 
-    // Since we're iterating, might as well prune here
-    for (const [key, entry] of this.map.entries()) {
-      if (entry.expirationTimestamp < now) {
-        this.map.delete(key) // In theory, this shouldn't happen frequently for feeds
+  get(key: string): T | undefined {
+    if (this.map.has(key)) {
+      const node = this.map.get(key) as LinkedListNode<ExpiringSortedSetEntry<T>>
+      const expired = node.data.expirationTimestamp <= Date.now()
+      if (expired) {
+        this.delete(key)
+        return undefined
       } else {
-        results.push(entry.value)
+        return node.data.value
       }
+    } else {
+      return undefined
     }
+  }
 
-    return results
+  delete(key: string): void {
+    if (this.map.has(key)) {
+      const node = this.map.get(key) as LinkedListNode<ExpiringSortedSetEntry<T>>
+      this.list.remove(node)
+      this.map.delete(key)
+    }
+  }
+
+  private moveToTail(node: LinkedListNode) {
+    this.list.remove(node)
+    this.list.insertAtTail(node)
+  }
+
+  private evictIfNeeded() {
+    if (this.list.size >= this.capacity) {
+      logger.warn(`List reached maximum capacity, evicting least recently updated entry`)
```

Comment:
Could we also log the item that's being removed here with the message? Think it could be useful when debugging.